src/expressions/method-call-expr.md

```diff
@@ -15,22 +15,155 @@ let log_pi = pi.unwrap_or(1.0).log(2.72);
-The following procedure is used:
 
-The first step is to build a list of candidate receiver types.
-Obtain these by repeatedly [dereferencing][dereference] the receiver expression's type, adding each type encountered to the list, then finally attempting an [unsized coercion] at the end, and adding the result type if that is successful.
-Then, for each candidate `T`, add `&T` and `&mut T` to the list immediately after `T`.
+The following procedure is used:
 
-For instance, if the receiver has type `Box<[i32;2]>`, then the candidate types will be `Box<[i32;2]>`, `&Box<[i32;2]>`, `&mut Box<[i32;2]>`, `[i32; 2]` (by dereferencing), `&[i32; 2]`, `&mut [i32; 2]`, `[i32]` (by unsized coercion), `&[i32]`, and finally `&mut [i32]`.
+## Determining candidate types
+
+First, a list of "candidate types" is assembled.
+
+These types are found by taking the receiver type and iterating, following either:
+
+* The built-in [dereference]; or
+* `<T as Receiver>::Target`
+
+to the next type. (If a step involved following the `Receiver` target, we also
+note whether it would have been reachable by following `<T as Deref::Target>`
+- this information is used later).
+
+At the end, an additional candidate step may be added for
+an [unsized coercion].
+
+Each step of this chain provides a possible `self` type for methods that
+might be called. The list will be used in two different ways:
+
+* To find types that might have methods. This is used in the
+  "determining candidate methods" step, described below. This considers
+  the full list.
+* To find types to which the receiver can be converted. This is used in the
+  "picking a method from the candidates" step, also described below - in this
+  case, we only consider the types reachable via `Deref` or built-in
+  dereferencing.
+
+There is a built-in implementation of `Receiver` for all `T: Deref`, so
+most of the time, every step can be reached through either mechanism.
+Sometimes, more types can be reached via the `Receiver` chain, and so
+more types will be considered for the former usage than the latter usage.
+
+For instance, if the receiver has type `Box<[i32;2]>`, then the candidate types
+will be `Box<[i32;2]>`,`[i32; 2]` (by dereferencing), and `[i32]` (by unsized
+coercion).
+
+If `SmartPtr<T>: Receiver<Target=T>`, and the receiver type is `&SmartPtr<Foo>`,
+then the candidate types would be `&SmartPtr<Foo>`, `SmartPtr<Foo>` and `Foo`.
+
+## Determining candidate methods
+
+This list of candidate types is then converted to a list of candidate methods.
+For each step, the `self` type is used to determine what searches to perform:
+
+* For a trait object, there is first a search for for inherent candidates for
+  the object, then inherent impl candidates for the type.
+* For a struct, enum, or foreign type, there is a search for inherent
+  impl candidates for the type.
+* For a type param, there's a search for for inherent candidates on the param.
```

Comment:
```suggestion
* For a type param, there's a search for inherent candidates on the param.
```

Reply:
Also, I'm unclear on what this means. There's no such thing as an inherent method on a type parameter. All I can think of is if there are `T: Trait` bounds on the type parameter, we treat any trait named in the bounds as inherent.

Reply:
Probably this is in reference to [these lines](https://github.com/rust-lang/rust/blob/9e48dfe5e0d5ec1b064b5d25504948ede0ec9b53/compiler/rustc_hir_typeck/src/method/probe.rs#L804-L806).

Reply:
As a bit of archaeology, it looks like @nikomatsakis added this in commit rust-lang/rust@8a8f200d102294cf1bd90cdacad995abccda7934 in 2012.

Reply:
This relates to the comment in https://github.com/rust-lang/reference/pull/1725/files#r1936405008.

The code uses "inherent" and "extension" candidates as terms of art.